bpf/headers/helpers.h

```diff
@@ -131,18 +150,91 @@ static inline int is_port_listen_udp_current_ns(void *ctx, __u32 ip, __u16 port)
     }
     return 0;
 }
+#endif
+
+#if ENABLE_IPV6
+static __u32 ip_zero6[4] = {0, 0, 0, 0};
+// ::1 (network order)
+static __u32 localhost6[4] = {0, 0, 0, 1 << 24};
+
+static inline int ipv6Equal(__u32 *a, __u32 *b)
+{
+    return a[0] == b[0] && a[1] == b[1] && a[2] == b[2] && a[3] == b[3];
+}
+
+static inline void setIpv6(__u32 *dst, __u32 *src)
+{
+    dst[0] = src[0];
+    dst[1] = src[1];
+    dst[2] = src[2];
+    dst[3] = src[3];
+}
+
+static inline int is_port_listen_current_ns6(void *ctx, __u32 *ip, __u16 port)
+{
+    struct bpf_sock_tuple tuple = {};
+    tuple.ipv6.dport = bpf_htons(port);
+    setIpv6(tuple.ipv6.daddr, ip);
+    struct bpf_sock *s = bpf_sk_lookup_tcp(ctx, &tuple, sizeof(tuple.ipv6),
+                                           BPF_F_CURRENT_NETNS, 0);
+    if (s) {
+        bpf_sk_release(s);
+        return 1;
+    }
+    return 0;
+}
+
+static inline int is_port_listen_udp_current_ns6(void *ctx, __u32 *ip,
+                                                 __u16 port)
+{
+    struct bpf_sock_tuple tuple = {};
+    tuple.ipv6.dport = bpf_htons(port);
+    setIpv6(tuple.ipv6.daddr, ip);
+    struct bpf_sock *s = bpf_sk_lookup_udp(ctx, &tuple, sizeof(tuple.ipv6),
+                                           BPF_F_CURRENT_NETNS, 0);
+    if (s) {
+        bpf_sk_release(s);
+        return 1;
+    }
+    return 0;
+}
+#endif
 
 struct origin_info {
     __u32 pid;
-    __u32 ip;
+    union {
+#if ENABLE_IPV4
+        __u32 ip4;
+#endif
+#if ENABLE_IPV6
+        __u32 ip6[4];
+#endif
+    };
     __u16 port;
     // last bit means that ip of process is detected.
     __u16 flags;
 };
 
 struct pair {
-    __u32 sip;
-    __u32 dip;
+    union {
+#if ENABLE_IPV4
+        __u32 sip4;
+#endif
+#if ENABLE_IPV6
```

Comment:
struct may not need to make a distinction based on these macros.
This not only affects understanding the code, but also makes the size of the map uncertain.

Reply:
The purpose is to reduce memory usage when enable ipv4 only (12B vs 48B)
If we don't need to care about the memory cost, I will remove the macros

Reply:
Most of the kernel code I've looked at doesn't distinguish between v4 and v6 by macros, so I don't think this is very necessary.
If you want to save memory, I think using union would be fine? What do you think?

Reply:
OK, removed